### ISSUE TITLE
kube-scheduler: Bring back the ability to stop at the first failed filter

### ISSUE THREAD

**Issue by towca** [kubernetes/kubernetes#114509]
#108829 changed the default behavior of scheduler framework's `RunFilterPlugins` - from stopping at the first failed filter to always running all filters. This change breaks Cluster Autoscaler at scale. Cluster Autoscaler imports the scheduler framework to simulate filters on artificial nodes. In a (pretty common) one-pod-per-node scenario, the number of `RunFilterPlugins` calls is ~quadratic wrt. the number of nodes needed for unschedulable pods - it can be a lot of calls. Before #108829, in such scenario the vast majority of `RunFilterPlugins` calls would stop at failing Fit or NodeAffinity filters, and not run the rest of them (e.g. PV-related filters). After the change, every call to `RunFilterPlugins` runs all filters, which completely breaks Cluster Autoscaler at scale (an example scenario of 300 1-pod-per-node pods and 200 node groups takes >10min in 1.24, and ~50s in <1.22).

Can we get the old behavior back? Or re-introduce the config option so that Cluster Autoscaler can set it accordingly? Ideally, we'd also like to get this backported as far as possible (the change was introduced in 1.24).

**Comment by Bryce-Soghigian** [comment]
I would be interested in adding a toggle flag that allows one to toggle on that behavior.

**Comment by Huang-Wei** [comment]
> https://github.com/kubernetes/kubernetes/pull/108829 changed the default behavior of scheduler framework's RunFilterPlugins - from stopping at the first failed filter to always running all filters.

That PR doesn't the behavior actually. Scheduler is always defaulted to stop upon the first Filter failure, and it's just we offered an option to continue go through other Filter plugins upon failures, and that option was removed recently.

Let me check the code to see if there is a regression.

**Comment by Huang-Wei** [comment]
When #108829 removed the following changes:

```diff
			if !f.runAllFilters {
				// Exit early if we don't need to run all filters.
				return statuses
			}
```

It should have kept `return statuses`. I will submit a fix along with test, and backport to 1.24, 1.25, 1.26. Thanks for reporting!

cc @ahg-g @kerthcet 

/assign
/kind bug
/kind regression
/priority critical-urgent
/triage/accepted

**Comment by ahg-g** [comment]
Thanks Kuba, this is indeed a regression on the scheduler end.

**Comment by ahg-g** [comment]
/open

for patches

**Comment by kerthcet** [comment]
Really sorry about this regression, and hope not to happen again. I think there's no e2e tests between kubernetes scheduler and cluster-autoscaler, so can we also have testcases of pods unschedulable in scheduling perf-test, maybe this can also help to prevent this happening again. 

UT benchmark test also makes sense, but it's not graphical.

**Comment by towca** [comment]
Thanks for the quick fix! Unfortunately I'm afraid this kind of issue would only be caught in a large-scale test scenario, which is not cheap to run continuously.

**Comment by Huang-Wei** [comment]
> so can we also have testcases of pods unschedulable in scheduling perf-test, maybe this can also help to prevent this happening again.

For this _particular_ issue, UT (added in the PR) is able to catch it. And hence doesn't need to be that complicated to test with cluster-autoscaler.

However, standing from a broader angle, why we didn't catch the regression in perf test is b/c the triggering actor is unschedulable Pods, instead of schedulable Pods (which are the happy path most perf tests covered). So going forward, I'd like to see more perf tests to mixin **unschedulable** Pods, and hence more likely to surface this kind of issue. (we have preemptio preemption test suite but doesn't seem adequate)

**Comment by kerthcet** [comment]
>  For this particular issue, UT (added in the PR) is able to catch it. And hence doesn't need to be that complicated to test with cluster-autoscaler.

This is not for the particular issue, but keeping an eye on whether there's any performance degradation between two kubernetes components, like we do for scheduling performance. but it's actually out of the scope of scheduler.

> I'd like to see more perf tests to mixin unschedulable Pods

I'll add it as one of the goals of https://github.com/kubernetes/kubernetes/issues/113586

**Comment by sanposhiho** [comment]
> However, standing from a broader angle, why we didn't catch the regression in perf test is b/c the triggering actor is unschedulable Pods, instead of schedulable Pods (which are the happy path most perf tests covered).

Actually, the scheduler_perf has a test case `Unschedulable`. And, the scheduler_perf collects the `scheduler_scheduling_attempt_duration_seconds` metric ([ref](https://github.com/kubernetes/kubernetes/blob/1edbb8cf1a864c2dd71a1e5b7282de4bc3d8f5fa/test/integration/scheduler_perf/scheduler_perf_test.go#L87)), which has the `result` label in it ([ref](https://github.com/kubernetes/kubernetes/blob/95bd687a284f63535cbf48b0696d8ae57c9929ef/pkg/scheduler/metrics/metrics.go#L65)). But, the problem is currently we don't distinguish the metric values by `result` label on each data; we just collect/aggregate it regardless the label. (We can distinguish the metric values by any labels in the scheduler_perf as we do in the `scheduler_framework_extension_point_duration_seconds` metric ([ref](https://github.com/kubernetes/kubernetes/blob/1edbb8cf1a864c2dd71a1e5b7282de4bc3d8f5fa/test/integration/scheduler_perf/scheduler_perf_test.go#L84-L85)))

I'd say just distinguishing the happy path and unhappy path with `result` label in `scheduler_scheduling_attempt_duration_seconds` metric should be enough to observe the particular performance degradation in the unhappy path.

**Comment by Huang-Wei** [comment]
@sanposhiho could you help share the perf-dash link where this metric (`scheduler_scheduling_attempt_duration_seconds`) is exposed?

**Comment by sanposhiho** [comment]
Hm, cannot see any scheduler metrics result from the current perf-dash 🤔 
I see some configuration that seems to be to parse the scheduler_perf result file though.

https://perf-dash.k8s.io/#/?jobname=scheduler-perf-benchmark&metriccategoryname=Scheduler&metricname=BenchmarkResults&benchmark=BenchmarkPerfScheduling%2FSchedulingBasic%2F5000Nodes-8&metricName=time

https://github.com/kubernetes/perf-tests/blob/f00bd48b94aec6c1c1a09e508bcece46f259a4bd/perfdash/config.go#L486-L494

TBH, I haven't visited the perf-dash so much, so not sure whether it originally doesn't work, it's not working for a long time, or got broken recently.

---

Also, another thing, I noticed the perf-dash doesn't show all test cases. All test cases after `SchedulingMigratedInTreePVs` haven't been executed.
https://github.com/kubernetes/kubernetes/blob/1edbb8cf1a864c2dd71a1e5b7282de4bc3d8f5fa/test/integration/scheduler_perf/config/performance-config.yaml#L99

**Comment by sanposhiho** [comment]
So... probably, something wrong happens in [`ci-benchmark-scheduler-perf-master`](https://github.com/kubernetes/test-infra/blob/de9e531d78cad06739c1409d14dffeb6de7052d7/config/jobs/kubernetes/sig-scalability/sig-scalability-periodic-jobs.yaml#L714) job (probably in the `SchedulingMigratedInTreePVs` test case) and the scheduler_perf is failing. The scheduler_perf doesn't make the result file when fails, so I'm guessing that's why we cannot see any scheduler metrics results from the perf-dash.

How can we check the result of the `ci-benchmark-scheduler-perf-master` job?

**Comment by liggitt** [comment]
cc @alculquicondor 

is this actually a regression, and is it resolved yet?

**Comment by Huang-Wei** [comment]
@liggitt It's resolved by #114518, and also cherry-picked to 1.24.x ~ 1.26.x.

/close
